File: 2395-longest-binary-subsequence-less-than-or-equal-to-k/longest-binary-subsequence-less-than-or-equal-to-k.java
class Solution {
    public int longestSubsequence(String s, int k) {
        int n = s.length();
        int weight = 0, p = 0;
        
        int result = 0;
        for (int i = n - 1; i >= 0; i--) {
            if ('1' == s.charAt(i)) {
                if (Math.pow(2, p) + weight <= k) {
                    result += 1;
                    weight += Math.pow(2, p);
                }
            } else {
                result += 1;
            }
            p += 1;
        }

        return result;
    }
}

File: 2395-longest-binary-subsequence-less-than-or-equal-to-k/longest-binary-subsequence-less-than-or-equal-to-k.cpp
class Solution {
public:
    int longestSubsequence(string s, int k) {
        int n = s.size();
        int weight = 0;

        int result = 0;
        for (int i = n - 1; i >= 0; i--) {
            if ('1' == s[i]) {
                if (std::pow(2, (n - 1) - i) + weight <= k) {
                    result += 1;
                    weight += pow(2, (n-1) - i);
                }
            } else {
                result += 1;
            }
        }

        return result;
    }
};